### PR TITLE
Add municipal boundaries layer

### LINKED ISSUE #34: Add Municipal Boundaries Layer
## Overview

- [ ] Find out which Municipal Boundaries layer we're using
- [ ] Decide if it can be queried from a remote source or if it needs to be ingested and tiled
  - [ ] If needed, ingest it
  - [ ] Tile it
- [ ] Add to map, wire up to Municipal Boundaries button under Basemap Layers

### PR DESCRIPTION
**Pull request by mstone121**
## Overview

This PR adds the municipal boundaries layer. The data is sourced from https://linc.osbm.nc.gov/explore/dataset/municipalities-2020/.

~The GeoJSON file is compressed with weback and then dynamically imported using the technique described here: https://www.azavea.com/blog/2022/04/28/how-to-code-split-geojson-layers-for-your-mapboxgl-js-react-apps/.~

The GeoJSON file is placed in the public folder and is retrieved with `fetch`.

Closes #34 

### Notes

The GeoJSON file compresses to about 4mb. If this is too much or the rendering of the entire file causes choppiness, then an alternative approach might be needed. A possibility is to split the layer into tiles and serve it from AWS.

## Testing Instructions

- Go to the draw page
- The municipal boundaries shoud show up.
- Clicking the sidebar visibility button should toggle the visibility of the layer

## Checklist

- [x] `fixup!` commits have been squashed
- [x] `CHANGELOG.md` updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [x] `README.md` updated if necessary to reflect the changes
- [x] CI passes after rebase

### CODE REVIEW

**Comment by rajadain** [comment]
Will take a look at this PR once it has been rebased to resolve conflicts

**Comment by rajadain** [comment]
Taking a look now

**Comment by rajadain** [comment]
Had to run `fetch-data` to get `update` to work correctly.

Upon running `server` and going to http://localhost:4545/draw, getting this out of memory error:

```
app_1         | <--- Last few GCs --->
app_1         |
app_1         | [36:0x3deb71f0]    26702 ms: Mark-sweep (reduce) 1953.6 (1990.1) -> 1953.6 (1989.3) MB, 276.2 / 0.0 ms  (average mu = 0.533, current mu = 0.001) last resort; GC in old space requested
app_1         | [36:0x3deb71f0]    26996 ms: Mark-sweep (reduce) 1953.6 (1989.3) -> 1953.6 (1989.3) MB, 293.0 / 0.0 ms  (average mu = 0.354, current mu = 0.000) last resort; GC in old space requested
app_1         |
app_1         |
app_1         | <--- JS stacktrace --->
app_1         |
app_1         | FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
app_1         |  1: 0xb4dbc8 node::Abort() [/usr/local/bin/node]
app_1         |  2: 0xa7b4ac node::FatalError(char const*, char const*) [/usr/local/bin/node]
app_1         |  3: 0xd037c0 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
app_1         |  4: 0xd0394c v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
app_1         |  5: 0xec3040 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]
app_1         |  6: 0xea5810 v8::internal::Factory::AllocateRaw(int, v8::internal::AllocationType, v8::internal::AllocationAlignment) [/usr/local/bin/node]
app_1         |  7: 0xe9caf8 v8::internal::FactoryBase<v8::internal::Factory>::AllocateRawArray(int, v8::internal::AllocationType) [/usr/local/bin/node]
app_1         |  8: 0xe9cc94 v8::internal::FactoryBase<v8::internal::Factory>::NewFixedArrayWithFiller(v8::internal::Handle<v8::internal::Map>, int, v8::internal::Handle<v8::internal::Oddball>, v8::internal::AllocationType) [/usr/local/bin/node]
app_1         |  9: 0x11335d4 v8::internal::MaybeHandle<v8::internal::OrderedHashMap> v8::internal::OrderedHashTable<v8::internal::OrderedHashMap, 2>::Allocate<v8::internal::Isolate>(v8::internal::Isolate*, int, v8::internal::AllocationType) [/usr/local/bin/node]
app_1         | 10: 0x11336b8 v8::internal::MaybeHandle<v8::internal::OrderedHashMap> v8::internal::OrderedHashTable<v8::internal::OrderedHashMap, 2>::Rehash<v8::internal::Isolate>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::OrderedHashMap>, int) [/usr/local/bin/node]
app_1         | 11: 0x122c724 v8::internal::Runtime_MapGrow(int, unsigned long*, v8::internal::Isolate*) [/usr/local/bin/node]
app_1         | 12: 0x160800c  [/usr/local/bin/node]
app_1         | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
app_1         | error Command failed with exit code 1.
iow-boundary-tool_app_1 exited with code 1
```

Investigating ...

**Comment by rajadain** [comment]
On `develop` the memory use of the `app` container stabilizes around ~500MB. On this branch it goes up to 2.2GB and then crashes. Going to see if increasing my Docker memory settings helps.

**Comment by rajadain** [comment]
I was able to get past the memory issues by [increasing the allocation](https://stackoverflow.com/questions/38558989/node-js-heap-out-of-memory) for node:

```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index 1452d04..fe4f703 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -22,7 +22,7 @@ services:
       - CHOKIDAR_USEPOLLING=true
       - CHOKIDAR_INTERVAL=100
       - PORT=4545
-      - NODE_OPTIONS=--openssl-legacy-provider
+      - NODE_OPTIONS=--openssl-legacy-provider --max_old_space_size=4096
     volumes:
       - ./src/app:/usr/local/src
       - ./src/django/static:/usr/local/src/build
```

I tried removing webpack-compression-plugin but that had no effect. It seems that as soon as `MunicipalBoundariesLayer` is imported, the memory use goes up from ~500MB to ~2.3GB. It touches ~3GB during the dev server startup.

File: .gitignore [hunk 1]
```diff
@@ -29,6 +29,7 @@ vendor.bundle.js
 bundle.js.map
 dist/
 yarn-error.log
+/src/app/src/layer_data
```
Comment (rajadain) [inline .gitignore:32]: Consider naming this folder `data` and adding a `.gitkeep` file so the folder is present in the repo, even if the data is not.

File: scripts/fetch-data [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then
+    if [ "${1:-}" = "--help" ]; then
+        usage
+    else
+        if [[ -f "$LAYER_DATA_DIR/$MUNICIPAL_BOUNDARIES_FILENAME" ]]; then
+            if [ "${1:-}" = "--force" ]; then
+                echo "Warning: forcing data refetching!"
+                rm "$LAYER_DATA_DIR/muni.gz"
```
Comment (rajadain) [inline scripts/fetch-data:36]: Should this be `muni.json`?

File: scripts/fetch-data [hunk 3]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+
+set -e
+
+if [[ -n "${IOW_BOUNDARY_TOOL_DEBUG}" ]]; then
+    set -x
+fi
+
+MUNICIPAL_BOUNDARIES_DATA_LOCATION="https://linc.osbm.nc.gov/explore/dataset/municipalities-2020/download/?format=geojson&lang=en"
+MUNICIPAL_BOUNDARIES_FILENAME=muni.json
```
Comment (rajadain) [inline scripts/fetch-data:10]: Consider naming this file `muni.geojson` to better communicate the kind of file it is (unless this leads to WebPack import issues?)

File: src/app/src/components/Basemaps.js [hunk 4]
```diff
@@ -31,16 +29,7 @@ export function SatelliteBasemap() {
 
 function Basemap({ url, attribution, maxZoomLevel }) {
     useSetMaxZoomLevel(maxZoomLevel);
-    const map = useMap();
+    useMapLayer(L.tileLayer(url, { attribution }));
```
Comment (rajadain) [inline src/app/src/components/Basemaps.js:32]: Nice

File: src/app/src/components/DrawMap/DrawMap.js [hunk 5]
```diff
@@ -8,11 +8,13 @@ import Map from '../Map';
 import EditToolbar from './EditToolbar';
 import MapControlButtons from './MapControlButtons';
 import MapFeatures from './MapFeatures';
+import MunicipalBoundariesLayer from './MunicipalBoundariesLayer';
```
Comment (rajadain) [inline src/app/src/components/DrawMap/DrawMap.js:11]: Consider moving this file out of `DrawMap` and into `components/Layers/MunicipalBoundaries.js`. We may also be able to move `Basemaps.js` in there.

**Review by rajadain — COMMENTED** [review]
I think the `--max-old-space-size=4096` workaround is good for this PR, but it won't work for the Parcel layer which is certainly more complex. In the interest of comparing and contrasting, I think let's go ahead with this GeoJSON approach for the Municipal Boundaries and use the mbtiles for the Parcel layer.

File: .gitignore (hunk 1 above)
Comment (mstone121) [reply .gitignore:32]: Good idea! What about using a `.gitignore` like this:
```.gitignore
# Ignore everything in this directory
*

# except this file
!.gitignore
```

Comment (rajadain) [reply .gitignore:32]: Ah. I like having a single root `.gitignore` that covers everything. Let's keep the entry there, and we can `git add --force src/app/src/data/.gitkeep` to add that one file while ignoring the rest.

File: scripts/fetch-data (hunk 3 above)
Comment (mstone121) [reply scripts/fetch-data:10]: It does cause WebPack import issues. What about `muni.geo.json`?

Comment (rajadain) [reply scripts/fetch-data:10]: Yes that sounds good!

**Comment by rajadain** [comment]
Let me know when this is ready for review again. Currently requests for `muni.geo.json` result in the main app being loaded again, which is probably 404 handling:

image

**Comment by caseycesari** [comment]
@rajadain You may need to run `./scripts/fetch-data` again to rename the file and put it in the new spot.

**Review by rajadain — APPROVED** [review]
>  need to run `./scripts/fetch-data` again

🤦 of course.

Did so, it works correctly now. Nice fixes, the code is very clean 👌 

Gates may have further updates to the styling of this layer, which is not very clear to observe, but we can make that adjustment in a follow up PR once it is available. This is good to merge!

Great job scripting the data import.

**Comment by rajadain** [comment]
This commit: [68f54b0](https://github.com/azavea/iow-boundary-tool/pull/40/commits/68f54b001b634aa3ce9dd2d8b42924d6b817880c) has a typo in the message, should be fixed.

> Add useMap~y~Layer hook

**Comment by rajadain** [comment]
Also confirmed that with the move of the JSON file to `public/data/` we no longer have the excess RAM usage, and no longer need the `--max-old-space-size=4096` flag:

image

**Pull request merged**